src/UserSpaceInstrumentation/InjectLibraryInTracee.cpp

```diff
@@ -0,0 +1,322 @@
+      .AppendImmediate64(address_dlsym)
+      .AppendBytes({0xff, 0xd0})
+      .AppendBytes({0xcc});
+  auto result_write_code = WriteTraceesMemory(pid, address_code, code.GetResultAsVector());
+  if (result_write_code.has_error()) {
+    FreeMemoryOrDie(pid, address_code, memory_size);
+    return result_write_code.error();
+  }
+
+  ExecuteOrDie(pid, original_registers, address_code);
+
+  void* return_value = absl::bit_cast<void*>(GetReturnValueOrDie(pid));
+
+  // Cleanup memory and registers.
+  RestoreRegistersOrDie(original_registers);
+  FreeMemoryOrDie(pid, address_code, memory_size);
+  return return_value;
+}
+
+[[nodiscard]] ErrorMessageOr<void> DlcloseInTracee(pid_t pid, void* handle) {
+  // Figure out address of dlclose.
+  OUTCOME_TRY(address_dlclose,
+              FindFunctionAddressWithFallback(pid, "dlclose", "libdl-", "__libc_dlclose", "libc-"));
+
+  // Backup registers.
+  RegisterState original_registers;
+  OUTCOME_TRY(original_registers.BackupRegisters(pid));
+
+  // Allocate small memory area in the tracee.
+  OUTCOME_TRY(address_code, AllocateInTracee(pid, kCodeScratchPadSize));
+
+  // We want to do the following in the tracee:
+  // dlclose(handle);
+  // The calling convention is to put the parameter in registers rdi. Then we load the address of
+  // address_dlclose into rax and do the call. Assembly in Intel syntax (destination first), machine
+  // code on the right:
+
+  // movabs rax, handle               48 b8 handle
+  // mov rdi, rax                     48 8b f8
+  // movabs rax, address_dlclose      48 b8 address_dlclose
+  // call rax                         ff d0
+  // int3                             cc
+  MachineCode code;
+  code.AppendBytes({0x48, 0xb8})
+      .AppendImmediate64(absl::bit_cast<uint64_t>(handle))
+      .AppendBytes({0x48, 0x8b, 0xf8})
+      .AppendBytes({0x48, 0xb8})
+      .AppendImmediate64(address_dlclose)
+      .AppendBytes({0xff, 0xd0})
+      .AppendBytes({0xcc});
+  auto result_write_code = WriteTraceesMemory(pid, address_code, code.GetResultAsVector());
+  if (result_write_code.has_error()) {
+    FreeMemoryOrDie(pid, address_code, kCodeScratchPadSize);
+    return result_write_code.error();
+  }
+
+  ExecuteOrDie(pid, original_registers, address_code);
+
+  if (GetReturnValueOrDie(pid) != 0) {
+    FATAL("Unable to unload dynamic library from tracee.");
+  }
+
+  // Cleanup memory and registers.
+  RestoreRegistersOrDie(original_registers);
+  FreeMemoryOrDie(pid, address_code, kCodeScratchPadSize);
+  return outcome::success();
+}
+
+ErrorMessageOr<uint64_t> FindFunctionAddress(pid_t pid, std::string_view function,
+                                             std::string_view module) {
+  auto modules = ReadModules(pid);
+  if (modules.has_error()) {
+    return modules.error();
+  }
+
+  std::string module_file_path;
+  uint64_t module_base_address = 0;
+  for (const auto& m : modules.value()) {
+    if (absl::StartsWith(m.name(), module)) {
```

Comment:
I still a bit worry about false positives here. "libc-core.so" for example. Maybe add a todo to replace it with regexp or keep iterating libraries if we have not found a function in this one?

Reply:
I have a fix for this - but it needs to go to ElfFile.cpp, you can merge it as is I will do a followup PR

Reply:
with #2070 module name will be set to soname which will let you to do exact comparison for module name with libdl.so and libc.so.

Reply:
Okay, I added a solution with a regex here already. I'll submit that and we can chat offline about  how to use module name.